# ProteomeGenerator

## Introduction

ProteomeGenerator is an open, modular, and scalable framework for reference guided and *de novo* proteogenomic database generation written in the [Snakemake](https://snakemake.readthedocs.io/en/stable/) workflow management system. The workflow consists of a base file that defines the project details and input samples, and pgm, a modular set of rules sourced as an include.

## Installation

### Dependencies

ProteomeGenerator requires [Miniconda](https://conda.io/miniconda.html) to install the [Snakemake](https://snakemake.readthedocs.io/en/stable/) workflow management system and, optionally, can be automatically containerized with [Singularity](https://singularity.lbl.gov/) in order to run in almost any Linux production environment. ProteomeGenerator depends on multiple free and open source tools. The following dependencies will be automatically downloaded and persistently installed within a Miniconda environment the first time you run the ProteomeGenerator workflow and loaded from cache thereafter.

* [STAR](https://github.com/alexdobin/STAR)
* [Picard](http://broadinstitute.github.io/picard/)
* [samtools](http://samtools.sourceforge.net)
* [StringTie](https://ccb.jhu.edu/software/stringtie/)
* [TransDecoder](https://github.com/TransDecoder/TransDecoder/wiki)
* [gffread](https://github.com/gpertea/gffread)
* [BLAST](https://blast.ncbi.nlm.nih.gov/Blast.cgi)
* [R](https://www.r-project.org/)
* [Bioconductor](https://bioconductor.org/)
* [Biostrings](https://bioconductor.org/packages/release/bioc/html/Biostrings.html)
* [mono](https://www.mono-project.com/)
* [MaxQuant](http://www.coxdocs.org/doku.php?id=maxquant:start)

Optionally, many of the outputs from ProteomeGenerator may be viewed using the [Integrative Genomics Viewer](http://software.broadinstitute.org/software/igv/).

### Installing ProteomeGenerator

ProteomeGenerator is installed to your preferred computing environment using git clone:

```bash
git clone https://github.com/jtpoirier/proteomegenerator
```

This command downloads Snakefile-K0562, the configuraton file used in Cifani et al., as well as the required includes and accessory scripts to run ProteomeGenerator on your data set.

## Configuration

To run ProteomeGenerator on your own data, create a new Snakefile using Snakefile-K0562 as a template. Next, edit the "User Variables" section to reflect your computing environment and samples. This section is broken into three parts: "Directories", "References", "Samples", and "MaxQuant".

### Directories

The "WD" variable sets the path that will serve as the relative path for all analyses. In a cluster environment, this path should be accessible to all nodes on the network. The "TMP" variable should point to a local scratch storage directory that should be mounted locally to each node on a cluster, as opposed to a high performance network storage location. This folder is not required to be accessible to all nodes. MaxQuant performs its calculations in this directory. Since MaxQuant is very read/write-intensive, local disk access should result in faster running times. Else, user may specify for another location.

### References

ProteomeGenerator requires a FASTA formated genome reference and accepts an optional GTF transcript model with coordinates based on the same reference. STAR and BLAST indexes are automatically generated by ProteomeGenerator based on the provided references. To perform *de novo* database generation, delete or comment out the "GTF" variable and ProteomeGenerator will adjust all necessary options on the fly. Regeneration of the STAR index is not required to replace or update the transcript model as long as the provided GTF is based on the same genome reference as the existing index.

### Samples

The "Samples" section should be edited to reflect the type, number, and naming structure of the input FASTQ files for a given experiment. In addition, the user must edit the input for the rule STAR to reflect the specific naming convention used.

### MaxQuant

The "MaxQuant" section has one variable, "RAW", that must be changed to reflect the directory containing all .raw files used in the MaxQuant pipeline. The other two variables should NOT be changed. MaxQuant requires a parameter file (generated by ProteomeGenerator or user-provided), FASTA file (generated by ProteomeGenerator), and Raw files (user-provided, same directory as WD).

The "MQ" variable points to MaxQuantCmd.exe, which is provided in "/MaxQuant/bin". This version of MaxQuant used is 1.6.2.3. You must use this version because the MaxQuant parameters file generated alters between versions. License agreement for redistribution in located in "/MaxQuant".

The "PAR" variable points to a template MaxQuant parameters file, which is provided in "/MaxQuant". This file is used to generate the actual parameters file used in MaxQuant.

Note: The generated MaxQuant parameters file will contain default settings to MaxQuant. To change these, you may create your own MaxQuant parameters file by executing "/MaxQuant/bin/MaxQuantGui.exe" on a Windows system. Click File > Load parameters... and load "/MaxQuant/mqpar_template.xml". Input all Raw data files. Adjust the parameters, including "Number of Processors". DO NOT add/remove FASTA files as this part will be generated by ProteomeGenerator. Finally, click File > Save parameters... and save it as "mqpar.xml" in the same directory as the template parameters. ProteomeGenerator will automatically detect this file and use it as input for MaxQuant instead of its self-generated parameters.


## Running ProteomeGenerator

ProteomeGenerator can be run locally or in a variety of high performance computing cluster environments. The below example demonstrates how to run ProteomeGenerator with Singularity on an [LSF](https://www.ibm.com/support/knowledgecenter/en/SSETD4/product_welcome_platform_lsf.html) cluster head node from within screen in case the connection to the head node is lost.



```bash
screen -S pg
snakemake --snakefile Snakefile-K0562 --cluster \
"bsub -J {params.J} -n {params.n} -R {params.R} -W 24:00 -o {params.o} -eo {params.eo}" \
--jn {rulename}.{jobid}.sj -j 50 -k --latency-wait 60 --use-conda --use-singularity --singularity-args "--bind /data:/data" --ri
```

"-W" wall time argument might need to be adjusted to account for larger datasets used in MaxQuant.

"--bind" argument will need to be adjusted so that Singularity can access scripts outside the container. You must bind the directory for ProteomeGenerator, "TMP", and "RAW" for MaxQuant so that ProteomeGenerator can read and write data on your system. Else, there will be an error associated with singularity.

### Expected Output

ProteomeGenerator will generate an indexed bam file of mapped and filtered reads of the format {sample}.Aligned.trimmed.out.bam, a sample-specific GTF of the format {sample}-stringtie.gtf, and a proteogenomic database called proteome.unique.fasta. A GFF3 corresponding to each entry in the fasta database is also generated with the predicted spliced peptide sequences mapped onto genome space for easy viewing in the [Integrative Genomics Viewer](http://software.broadinstitute.org/software/igv/). If a reference GTF file is provided, the pipeline will generate a proteogenomic database based on these reference annotations in addition to the sample-specific database.

Furthermore, it will generate all tables outputted from MaxQuant in the user-specified working directory, under "out/combined/txt/".

## Citing ProteomeGenerator

Please cite [https://www.biorxiv.org/content/early/2017/12/19/236844](https://www.biorxiv.org/content/early/2017/12/19/236844)
