---
title: "K052 Test"
output:
  html_document:
    smart: false
---

```{r init, echo=FALSE, message=FALSE, warning=FALSE, error=TRUE, cache=FALSE}
library(knitr)
#opts_knit$set(root.dir = snakemake@workdir)
library(rtracklayer)
library(Biostrings)
library(GenomicFeatures)
library(Rsamtools)
library(GenomicAlignments)
library(VariantAnnotation)
library(ggbio)
```

## Custom Reference

Compares reference genome .fasta to custom genome .fasta. Missense mutation occurs at the 6th amino acid shown below.

```{r fasta, echo=FALSE, message=FALSE, warning=FALSE, error=TRUE, cache=FALSE}
ref <- readDNAStringSet(snakemake@input[["ref"]])
ref <- ref[[1]]
ref <- ref[2782:2826]
cat("Reference DNA Sequence [2782:2826]:\n")
as.character(ref)
cus <- readDNAStringSet(snakemake@input[["custom"]])
cus <- cus[[1]]
cus <- cus[2782:2826]
cat("Custom DNA Sequence [2782:2826]:\n")
as.character(cus)

refaa <- reverseComplement(ref)
refaa <- translate(refaa)
cat("Reference AA sequence:\n")
as.character(refaa)
cusaa <- reverseComplement(cus)
cusaa <- translate(cusaa)
cat("Custom AA sequence:\n")
as.character(cusaa)
```

## Test File Overview

Plots the reads, coverage, and variant annotations of small K052 test dataset.


```{r variants, echo=FALSE, message=FALSE, warning=FALSE, error=TRUE, cache=FALSE}
ga <- readGAlignments(snakemake@input[["bam"]], use.names=TRUE)
rd <- autoplot(ga, geom = "rect") + xlim(2000, 3000)
cov <- autoplot(ga, geom = "line", stat = "coverage") + xlim(2000, 3000)
vcf <- readVcf(file=snakemake@input[["vcf"]])
vcf <- autoplot(vcf, type="fixed") + xlim(2000, 3000)
tracks(Reads=rd, Coverage=cov, Variants=vcf, heights = c(0.2, 0.2, 0.2)) + ylab("")
```